packages/protocol/contracts/L1/libs/LibProving.sol

```diff
@@ -22,23 +22,91 @@ library LibProving {
         address prover
     );
 
-    event ConflictingProof(
-        uint64 blockId,
-        bytes32 parentHash,
-        bytes32 conflictingBlockHash,
-        bytes32 conflictingSignalRoot,
-        bytes32 blockHash,
-        bytes32 signalRoot
-    );
-
     error L1_ALREADY_PROVEN();
     error L1_BLOCK_ID();
     error L1_EVIDENCE_MISMATCH(bytes32 expected, bytes32 actual);
     error L1_FORK_CHOICE_NOT_FOUND();
     error L1_INVALID_PROOF();
     error L1_INVALID_EVIDENCE();
+    error L1_INVALID_ORACLE();
+    error L1_ORACLE_DISABLED();
+    error L1_NOT_ORACLE_PROVEN();
     error L1_NOT_ORACLE_PROVER();
     error L1_UNEXPECTED_FORK_CHOICE_ID();
+    error L1_CONFLICTING_PROOF(
+        uint256 id,
+        bytes32 parentHash,
+        bytes32 conflictingBlockHash,
+        bytes32 conflictingSignalRoot,
+        bytes32 blockHash,
+        bytes32 signalRoot
+    );
+
+    function oracleProveBlocks(
+        TaikoData.State storage state,
+        TaikoData.Config memory config,
+        AddressResolver resolver,
+        uint256 blockId,
+        TaikoData.BlockOracle[] memory oracles
+    ) internal {
+        if (!config.enableOracleProver) revert L1_ORACLE_DISABLED();
+        if (msg.sender != resolver.resolve("oracle_prover", false))
+            revert L1_NOT_ORACLE_PROVER();
+
+        TaikoData.ForkChoice storage fc;
+        for (uint i = 0; i < oracles.length; ) {
+            TaikoData.BlockOracle memory oracle = oracles[i];
+            uint256 id = blockId + i;
+
+            if (id <= state.lastVerifiedBlockId || id >= state.numBlocks)
+                revert L1_BLOCK_ID();
+
+            if (
+                oracle.parentHash == 0 ||
+                oracle.blockHash == 0 ||
+                oracle.blockHash == oracle.parentHash ||
+                oracle.signalRoot == 0
+            ) revert L1_INVALID_ORACLE();
+
+            TaikoData.Block storage blk = state.blocks[
+                id % config.ringBufferSize
+            ];
+
+            uint256 fcId = state.forkChoiceIds[id][oracle.parentHash];
+            if (fcId == 0) {
+                fcId = blk.nextForkChoiceId;
+                unchecked {
+                    ++blk.nextForkChoiceId;
+                }
+                assert(fcId > 0);
+
+                fc = blk.forkChoices[fcId];
+                state.forkChoiceIds[id][oracle.parentHash] = fcId;
+            } else {
+                fc = blk.forkChoices[fcId];
+                if (fc.prover != address(0)) revert L1_ALREADY_PROVEN();
```

Comment:
Maybe still allow overwriting even when already proven by another prover? Only possible when ZKPs are incomplete/buggy, but we're likely to need something that'll allow us to do this somehow.